### PR TITLE
Disable ptrace on arm64e running iOS 14.2 and greater

### PR DESCRIPTION
**Pull request by jmagman**
## Description

The ptrace trick may not be necessary on iOS 14.2 (beta release candidate) which seems to have added [JIT compilation support](https://9to5mac.com/2020/11/06/ios-14-2-brings-jit-compilation-support-which-enables-emulation-apps-at-full-performance/).  This means Flutter debug apps aren't [sigkilled when launched without an attached debugger](https://github.com/flutter/engine/pull/20980).

There's [speculation](https://github.com/utmapp/UTM/issues/516#issuecomment-723293451) that this is only true on A12 (and presumably later A-) chips.  With `#define TRACING_CHECKS_NECESSARY 0` a debug app launched from the home screen on my A12 iPad Pro, but on @xster's A9 iPhone 6s it still crashed.

https://github.com/flutter/engine/blob/f0fb74b8fb52909e5bce45e176b00db1f0acae9e/runtime/ptrace_check.h#L13-L15

The outstanding theory is [JIT may only be working on arm64e devices](https://github.com/flutter/flutter/issues/69984#issuecomment-723363587) running iOS 14.2.  

## Related Issues

Fixes https://github.com/flutter/flutter/issues/69984

## Testing
I have no idea how to test this any more than the rest of this `ptrace` file without adding some wrappers around `sysctlbyname`.  Unfortunately I don't think we have any [arm64e devicelab test devices](https://github.com/flutter/flutter/issues/70004), either.

### CODE REVIEW

**Comment by jmagman** [comment]
Let's hold off actually merging this until we know more about what devices still need tracing, but I wanted a PR ready to go once in case.

File: runtime/ptrace_check.cc
Comment (jmagman) [inline runtime/ptrace_check.cc:159]: This parsing felt like an interview question (`atoi`!), happy to try something more idiomatic if ya'll can think of a better way to do it.  I started with an `istringstream` but it actually wound up being identical logic so it didn't seem worth the extra object.

File: runtime/ptrace_check.cc
Comment (xster) [inline runtime/ptrace_check.cc:143]: up to you but it feels like we're building in a lot of platform specific knowledge at this point. How about just letting this file statically take a SetOSVersionMayNeedTracingCheck injection or something and FlutterEngine can just set it to https://developer.apple.com/documentation/foundation/nsprocessinfo/1414876-isoperatingsystematleastversion? Could save us some work.

File: runtime/ptrace_check.cc
Comment (xster) [inline runtime/ptrace_check.cc:140]: ultra-nit: there's a triple negation at this point :) false to != arm64 to unnecessary. Can we simplify slightly?

File: runtime/ptrace_check.cc
Comment (jmagman) [reply runtime/ptrace_check.cc:143]: If this were all in obj-c++ world then I wouldn't have bothered with all these syctl checks.  But it's being injected from:
https://github.com/flutter/engine/blob/e78e405e6a2fcb79956280be019e916a12a8de05/runtime/dart_vm.cc#L333

Comment (xster) [reply runtime/ptrace_check.cc:143]: that's what I meant injecting it statically (since all these stuff is static anyway). Since the engine calls first, we know it would be set before dart calls it again.

Comment (jmagman) [reply runtime/ptrace_check.cc:143]: I don't love that the lower level code would require a static variable to be set before it can be called, without enforcing that ordering itself.  Seems like an easy opportunity for bugs if this ever gets refactored.  Will defer to @chinmaygarde's opinion here.

If we decide to keep this `kern.osversion` parsing I'll add a unit test.

File: runtime/ptrace_check.cc
Comment (xster) [inline runtime/ptrace_check.cc:30]: Should this be conditionally imported and used? @chinmaygarde might have a better idea of whether this is available on all our target platforms.

**Review by xster — APPROVED** [review]

File: runtime/ptrace_check.cc
Comment (chinmaygarde) [reply runtime/ptrace_check.cc:30]: This is fine. `TRACING_CHECKS_NECESSARY` is unset on all other platforms.

File: runtime/ptrace_check.cc
Comment (chinmaygarde) [inline runtime/ptrace_check.cc:127]: Here and elsewhere: Early return in case of error.

File: runtime/ptrace_check.cc
Comment (chinmaygarde) [reply runtime/ptrace_check.cc:143]: I love the initiative in doing this in C, but let's make this an ObjC++ TU and use `NSOperatingSystemVersion` instead. You can remove the `TRACING_CHECKS_NECESSARY` macro and add a stub C++ file for non-Darwin platforms.

File: runtime/ptrace_check.cc
Comment (chinmaygarde) [inline runtime/ptrace_check.cc:145]: Unchecked error code. Though, I'd suggest using the ObjC++ variant instead.

**Review by chinmaygarde — CHANGES_REQUESTED** [review]
Construction issues aside, this must not land as you'd still to check if the entitlement is present at the specific iOS version. Apps without the entitlement in debug will still crash. I'd also rather link to Apple's documentation instead of relying on speculation right now.

**Comment by chinmaygarde** [comment]
@jmagman Any updates? Have you tried this without the entitlement in the plist?

**Comment by jmagman** [comment]
> @jmagman Any updates? 

I haven't created the ObjC++ variant yet.  On this PR it's working on an (arm64e) iPad Pro by launching without the debugger attached and running the app.  And on an (arm64) iPhone 6s it's showing the warning screen.

>Have you tried this without the entitlement in the plist?
> you'd still to check if the entitlement is present at the specific iOS version.

Which entitlement?  `get-task-allow`?

> I'd also rather link to Apple's documentation instead of relying on speculation right now.

Yeah all I have is speculation about the A12 and higher chips.

**Comment by jmagman** [comment]
@zanderso Would you like me to keep working on this so users can launch apps from the home screen on new devices?  Or should I abandon?

**Comment by zanderso** [comment]
I think we should probably abandon until we can get some more solid info about the devices where this is supported.

**Pull request closed without merging**
